src/icalendar/cal/event.py

```diff
@@ -368,6 +392,112 @@ def end(self, end: date | datetime | None):
         """Set the end."""
         self.DTEND = end
 
+    def set_duration(self, duration: timedelta | None, locked: str = "start"):
+        """Set the duration with explicit locking behavior.
+
+        Args:
+            duration: The duration to set, or None to convert to DURATION property
+            locked: Which property to keep unchanged ('start' or 'end')
+        """
+        # Convert to DURATION property if duration is None
+        if duration is None:
+            if "DURATION" in self:
+                return  # Already has DURATION property
+            current_duration = self.duration
+            self.pop("DTEND", None)
+            self.add("DURATION", current_duration)
+            return
+
+        if not isinstance(duration, timedelta):
+            raise TypeError(f"Use timedelta, not {type(duration).__name__}.")
+
+        # Validate date/duration compatibility
+        start = self.DTSTART
+        if (
+            isinstance(start, date)
```

Comment:
There is a shortcut `icalendar.util.is_date()`

Reply:
Also, `isinstance(start, date)` may not be what you want ... the standart datetime is a subclass of date if I'm not mistaken.

Reply:
(sorry for the noise, I see you take care of this on the next line) :-)